# Odds

One of the difficulties with bayesian stats is that we need to divide by a normalisation constant, the total likelihood.

$$ posterior = \frac{likelihood * prior}{total likelihood}$$

That total likelihood constant can be difficult to calculate. But, if we use odds then we can eliminate it from calculations.

$$odds = \frac{P(X)}{P(¬X)}$$

$$probability = \frac{odds}{odds + 1}$$

As an example, if there is a game that we have 10% chance of winning then the odds are 2:1...

In [1]:
def Odds(x_for, x_against):
    return x_for / x_against

def Probability(odds):
    return odds / (odds + 1)

In [2]:
od = Odds(1, 2)

od

0.5

In [3]:
Probability(od)

0.3333333333333333

Note that the odds are 2:1 which ends up being 0.5 i.e. the change we win is half the size of the chance to lose but but the probability of winning is only 1/3. (Out of all possible outcomes, we only win in one outcome.)

## Odds in Bayes Theorem

As I noted above, we can use a fraction of two probabilities to simplify bayes theorem.

$$\frac{P(A|D)}{P(B|D)} = \frac{P(A) * P(D|A)}{P(B) * P(D|B)} $$

If A and B are mutually exclusive and collectively exhaustive then the odds are calculated as such...

$$o(A|D) = o(A) \frac{p(D|A)}{p(D|B)}$$

## Example: Cookie Problem

Suppose again that there are two powls of cookies:

bowl | vanilla | chocolate
--- | --- | ---
1 | 30 | 10
2 | 20 |20

If we choose one cookie that is vanilla what is the probability that it came from bowl 1?

In [4]:
# Prior prob of choosing 1 is 0.5
o_A = 0.5 / (1-0.5)

o_A  #i.e. 1:1

1.0

In [5]:
p_D_giv_A = 3/4
p_D_giv_B = 2/4

# Therefore o(A|D) is...
o_A * (p_D_giv_A / p_D_giv_B)

1.5

The above answer is 1.5:1 odds, otherwise written as 3:2. 

This probability corresponds to $3/(3+2) = 3/5$